sign/sign.go

```diff
@@ -0,0 +1,103 @@
+// Package sign provides unified interfaces for signature schemes.
+package sign
+
+import (
+	"crypto"
+	"encoding"
+	"errors"
+)
+
+type SignatureOpts struct {
+	// If non-empty, includes the given context in the signature if supported
+	// and will cause an error during signing otherwise.
+	Context string
+}
+
+// A public key is used to verify a signature set by the corresponding private
+// key.
+type PublicKey interface {
+	// Returns the signature scheme for this public key.
+	Scheme() Scheme
+	Equal(crypto.PublicKey) bool
+	encoding.BinaryMarshaler
+	crypto.PublicKey
+}
+
+// A private key allows one to create signatures.
+type PrivateKey interface {
+	// Returns the signature scheme for this private key.
+	Scheme() Scheme
+
+	Equal(crypto.PrivateKey) bool
+	// For compatibility with Go standard library
+	crypto.Signer
+	crypto.PrivateKey
+	encoding.BinaryMarshaler
+}
+
+// A Scheme represents a specific instance of a signature scheme.
+type Scheme interface {
+	// Name of the scheme
+	Name() string
+
+	// GenerateKey creates a new key-pair.
+	GenerateKey() (PublicKey, PrivateKey, error)
+
+	// Creates a signature using the PrivateKey on the given message and
+	// returns the signature. opts are additional options which can be nil.
+	Sign(sk PrivateKey, message []byte, opts *SignatureOpts) []byte
+
+	// Checks whether the given signature is a valid signature set by
+	// the private key corresponding to the given public key on the
+	// given message. opts are additional options which can be nil.
+	Verify(pk PublicKey, message []byte, signature []byte, opts *SignatureOpts) bool
+
+	// Deterministically derives a keypair from a seed.  If you're unsure,
+	// you're better off using GenerateKey().
+	//
+	// Panics if seed is not of length SeedSize().
+	DeriveKey(seed []byte) (PublicKey, PrivateKey)
+
+	// Unmarshals a PublicKey from the provided buffer.
+	UnmarshalBinaryPublicKey([]byte) (PublicKey, error)
+
+	// Unmarshals a PublicKey from the provided buffer.
+	UnmarshalBinaryPrivateKey([]byte) (PrivateKey, error)
+
+	// Size of binary marshalled public keys
+	PublicKeySize() int
+
+	// Size of binary marshalled public keys
+	PrivateKeySize() int
+
+	// Size of signatures
+	SignatureSize() int
+
+	// Size of seeds
+	SeedSize() int
+
+	// Returns whether contexts are supported
+	SupportsContext() bool
```

Comment:
I am not sure whether this method is required. As per the policy mentioned above, if context is provided and not required then it should fail.

Reply:
It's nice to know beforehand whether a scheme will by useable when you require a context.  (I need this in the tests to check whether to provide a context or not.)

Reply:
Is this the only use case? I think this issue was one of the reasons of the previous policy. A scheme uses the context if the scheme supports it, otherwise it is simply ignored. 
Anyway, either policy works.

Reply:
Ignoring context if the scheme doesn't support it is very dangerous: a protocol might essentially require different signatures for different contexts.  If we ignore the context for schemes that don't support it, we leave a foot gun.

Reply:
good point.